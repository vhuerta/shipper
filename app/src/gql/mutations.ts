import { useMutation, useQuery } from '@apollo/client';
import gql from 'graphql-tag';

type PackagesData = {
  result: {
    packages: Package[];
  };
};

type PackagesVariables = {
  packages: Package[];
};

export const PACK_ITEMS = gql`
  mutation PackItems($packages: [PackageInput]!) {
    result: pack_items(data: { packages: $packages }) {
      packages {
        id
        line_items {
          id
          quantity
          sku
          location
        }
      }
    }
  }
`;

export const usePackItems = () => useMutation<PackagesData, PackagesVariables>(PACK_ITEMS);
