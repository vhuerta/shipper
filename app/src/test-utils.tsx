import { ThemeProvider } from '@emotion/react';
import { render, RenderOptions } from '@testing-library/react-native';
import React from 'react';
import { StoreProvider } from './store/Store';
import { theme } from './theming/theme';

type AllTheProvidersProps = {
  children: JSX.Element;
};

function AllTheProviders({ children }: AllTheProvidersProps) {
  return (
    <ThemeProvider theme={theme}>
      <StoreProvider>{children}</StoreProvider>
    </ThemeProvider>
  );
}

const customRender = (
  component: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
  options?: RenderOptions | undefined
) => render(component, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { customRender as render };
