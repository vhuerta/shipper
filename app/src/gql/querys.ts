import { useQuery } from '@apollo/client';
import gql from 'graphql-tag';

type ItemsData = {
  items: Item[];
};

export const FETCH_ITEMS = gql`
  query FetchItems {
    items: line_items {
      id
      quantity
      sku
      location
    }
  }
`;

export const useItemsQuery = () => useQuery<ItemsData>(FETCH_ITEMS);
