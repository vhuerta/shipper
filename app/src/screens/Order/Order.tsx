import { useNavigation } from '@react-navigation/core';
import { StackNavigationProp } from '@react-navigation/stack';
import React, { useEffect } from 'react';
import { MainContainer, TitleContainer } from '../../components/Layout.styled';
import { Subtitle, SubtitleBold, Title } from '../../components/Typography.styled';
import { useItemsQuery } from '../../gql/querys';
import { MainNavigationParamList } from '../../routes/MainNavigation';
import { addItem, resetOrder } from '../../store/actions';
import { countActivePackageItems, countTotalItems } from '../../store/selectors';
import { useStore } from '../../store/Store';
import ItemsList from './components/ItemsList';
import OrderButton from './components/OrderButton';

export default function Order() {
  const navigation = useNavigation<StackNavigationProp<MainNavigationParamList>>();
  const { state, dispatch } = useStore();
  const { data } = useItemsQuery();
  const items = data?.items || [];

  useEffect(() => {
    dispatch(resetOrder());
  }, []);

  function handleShowOrderDetail() {
    navigation.navigate('OrderDetail');
  }

  function handleTapItem(item: Item) {
    dispatch(addItem(item));
  }

  return (
    <MainContainer edges={['top']}>
      <TitleContainer>
        <Title>Products</Title>
        <OrderButton
          itemsCount={countTotalItems(state)}
          onPress={handleShowOrderDetail}
        />
      </TitleContainer>
      <Subtitle>
        <SubtitleBold>Active Package: {state.activePackage}</SubtitleBold>
        {'  |  '}
        <SubtitleBold>Packed Products: {countActivePackageItems(state)}</SubtitleBold>
      </Subtitle>
      <Subtitle>Tap on a product to add it to the package.</Subtitle>
      <ItemsList items={items} onPressItem={handleTapItem} />
    </MainContainer>
  );
}
