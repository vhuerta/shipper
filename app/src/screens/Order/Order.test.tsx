import { useNavigation } from '@react-navigation/core';
import { fireEvent, waitFor } from '@testing-library/react-native';
import React from 'react';
import { useItemsQuery } from '../../gql/querys';
import { render } from '../../test-utils';
import Order from './Order';

const mockedNavigate = jest.fn();

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

jest.mock('@react-navigation/core', () => ({
  useNavigation: () => ({
    navigate: mockedNavigate,
  }),
}));

jest.mock('../../gql/querys', () => ({
  useItemsQuery: jest.fn().mockReturnValue({ data: { items: undefined } }),
}));

describe('<Order />', () => {
  test('renders correctly without items', () => {
    const tree = render(<Order />);
    expect(tree).toMatchSnapshot();
  });

  test('pressing the order button triggers navigation to the order datail view', () => {
    const { getByTestId, debug } = render(<Order />);
    fireEvent(getByTestId('order-button'), 'press');
    expect(mockedNavigate.mock.calls[0][0]).toEqual('OrderDetail');
  });

  test('renders item list correctly and adds tap over each item and the item get added to the package', async () => {
    const items = [
      {
        id: 1,
        sku: 'product-1',
        location: 'a1',
        quantity: 1,
      },
      {
        id: 2,
        sku: 'product-2',
        location: 'a2',
        quantity: 1,
      },
    ];

    (useItemsQuery as jest.Mock).mockReturnValue({ data: { items } });

    const { getByText, getByTestId } = render(<Order />);
    getByText('Packed Products: 0');
    getByText('Active Package: 1');

    getByText(`SKU: ${items[0].sku}`);
    fireEvent(getByTestId(`item-${items[0].id}`), 'press');
    await waitFor(() => expect(getByText(`Packed Products: 1`)).toBeTruthy());

    getByText(`SKU: ${items[0].sku}`);
    fireEvent(getByTestId(`item-${items[1].id}`), 'press');
    await waitFor(() => expect(getByText(`Packed Products: 2`)).toBeTruthy());
  });
});
