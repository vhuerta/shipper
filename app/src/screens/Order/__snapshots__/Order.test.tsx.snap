// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Order /> renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "top",
    ]
  }
  style={
    Array [
      Object {
        "backgroundColor": "#1c2128",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingBottom": 0,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 20,
          "marginTop": 20,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#eaebf9",
            "fontSize": 32,
          },
          undefined,
        ]
      }
    >
      Products
    </Text>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 45,
        }
      }
      testID="order-button"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#00b589",
              "borderBottomLeftRadius": 18,
              "borderBottomRightRadius": 18,
              "borderTopLeftRadius": 18,
              "borderTopRightRadius": 18,
              "height": 18,
              "justifyContent": "center",
              "minWidth": 18,
              "paddingBottom": 0,
              "paddingLeft": 5,
              "paddingRight": 5,
              "paddingTop": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": 1,
            },
            undefined,
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#eaebf9",
                "fontSize": 12,
              },
              undefined,
            ]
          }
        >
          0
        </Text>
      </View>
      <Text />
    </View>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#eaebf9",
          "fontSize": 14,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#00b589",
            "fontSize": 14,
            "fontWeight": "bold",
          },
          undefined,
        ]
      }
    >
      Active Package: 
      1
    </Text>
      |  
    <Text
      style={
        Array [
          Object {
            "color": "#00b589",
            "fontSize": 14,
            "fontWeight": "bold",
          },
          undefined,
        ]
      }
    >
      Packed Products: 
      0
    </Text>
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#eaebf9",
          "fontSize": 14,
        },
        undefined,
      ]
    }
  >
    Tap on a product to add it to the package.
  </Text>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "marginTop": 10,
        },
        undefined,
      ]
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</RNCSafeAreaView>
`;

exports[`<Order /> renders correctly without items 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "top",
    ]
  }
  style={
    Array [
      Object {
        "backgroundColor": "#1c2128",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingBottom": 0,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 20,
          "marginTop": 20,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#eaebf9",
            "fontSize": 32,
          },
          undefined,
        ]
      }
    >
      Products
    </Text>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 45,
        }
      }
      testID="order-button"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#00b589",
              "borderBottomLeftRadius": 18,
              "borderBottomRightRadius": 18,
              "borderTopLeftRadius": 18,
              "borderTopRightRadius": 18,
              "height": 18,
              "justifyContent": "center",
              "minWidth": 18,
              "paddingBottom": 0,
              "paddingLeft": 5,
              "paddingRight": 5,
              "paddingTop": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": 1,
            },
            undefined,
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#eaebf9",
                "fontSize": 12,
              },
              undefined,
            ]
          }
        >
          0
        </Text>
      </View>
      <Text />
    </View>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#eaebf9",
          "fontSize": 14,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#00b589",
            "fontSize": 14,
            "fontWeight": "bold",
          },
          undefined,
        ]
      }
    >
      Active Package: 
      1
    </Text>
      |  
    <Text
      style={
        Array [
          Object {
            "color": "#00b589",
            "fontSize": 14,
            "fontWeight": "bold",
          },
          undefined,
        ]
      }
    >
      Packed Products: 
      0
    </Text>
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#eaebf9",
          "fontSize": 14,
        },
        undefined,
      ]
    }
  >
    Tap on a product to add it to the package.
  </Text>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "marginTop": 10,
        },
        undefined,
      ]
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</RNCSafeAreaView>
`;
